File: src/main/golang/leetcode/editor/cn/[279]Perfect Squares.go
package lc

import "math"

//leetcode submit region begin(Prohibit modification and deletion)
func numSquares(n int) int {
	Min := func(x, y int) int {
		if x > y {
			return y
		}
		return x
	}
	dp := make([]int, n+1)
	for i := 0; i < n+1; i++ {
		dp[i] = math.MaxInt32
	}
	dp[0] = 0
	for i := 1; i <= 100; i++ {
		for j := i * i; j <= n; j++ {
			dp[j] = Min(dp[j], dp[j-i*i]+1)
		}
	}
	return dp[n]
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[279]Perfect Squares.py
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def numSquares(self, n: int) -> int:
        square = [i * i for i in range(1, int(n ** 0.5) + 1)]
        dp = [0] * (n + 1)
        for i in range(1, n + 1):
            dp[i] = min([dp[i - j] for j in square if j <= i]) + 1
        return dp[n]
# leetcode submit region end(Prohibit modification and deletion)
